feat(linter): implement prefer-const

connorshea opened the pull request:
Taking a stab at implementing prefer-const, but I could probably use someone else's help to finish this up.

This was primarily implemented on a whim a few weeks ago, using GitHub Copilot and Claude Sonnet 4.5. I've gotten all the tests passing, so I think this is ready for review. Unfortunately as a Rust noob I have little ability to tell whether this code is following best practices (it almost certainly isn't, it feels a bit janky even as someone unfamiliar with Rust). I've done some refactoring based on code smells I've noticed, but it's still pretty smelly.

I did test it on the mastodon repo as well to ensure it detects problematic code correctly, and it gets identical results to what the eslint rule reports: https://github.com/oxc-project/oxc/pull/15707#issuecomment-3537083686

Fixes #12645.

connorshea committed:
First pass at implementing the prefer-const rule.

Entirely done by GitHub Copilot w/ Claude Sonnet 4.5.

connorshea committed:
Iteration 2, implements some of the options.

connorshea committed:
Enable a few more tests.

connorshea committed:
Enable a few more tests.

connorshea committed:
Enable more tests.

connorshea committed:
Get a few more tests passing.

connorshea committed:
Format and update snap.

connorshea committed:
Implement more of the ignoreReadBeforeAssign option.

connorshea committed:
Get more tests passing!

connorshea committed:
Get more tests passing!

connorshea committed:
Fix a few more tests related to control flow.

connorshea committed:
Update a weird test.

Unclear why this was done this way in eslint's tests originally.

connorshea committed:
Get another test passing.

connorshea committed:
This is becoming very complex and I think the AI is doing excessive things here now...

connorshea committed:
Re-run lintgen.

connorshea committed:
Fix lint failures.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

camc314 commented on crates/oxc_linter/src/rules/eslint/prefer_const.rs:
```diff
@@ -0,0 +1,1033 @@
+        // (
+        //     "let { name, ...otherStuff } = obj; otherStuff = {};",
+        //     Some(serde_json::json!([{ "destructuring": "all" }])),
+        // ), // { "ecmaVersion": 2018 },
```
lets fix this in the ast impl:
```patch
diff --git a/crates/oxc_ast/src/ast_impl/js.rs b/crates/oxc_ast/src/ast_impl/js.rs
index b98cd035d..2d331aa7e 100644
--- a/crates/oxc_ast/src/ast_impl/js.rs
+++ b/crates/oxc_ast/src/ast_impl/js.rs
@@ -1326,9 +1326,14 @@ impl<'a> BindingPatternKind<'a> {
                 .iter()
                 .filter_map(|item| item.as_ref())
                 .for_each(|item| item.kind.append_binding_identifiers(idents)),
-            Self::ObjectPattern(pattern) => pattern.properties.iter().for_each(|item| {
-                item.value.kind.append_binding_identifiers(idents);
-            }),
+            Self::ObjectPattern(pattern) => {
+                pattern.properties.iter().for_each(|item| {
+                    item.value.kind.append_binding_identifiers(idents);
+                });
+                if let Some(rest) = &pattern.rest {
+                    rest.argument.kind.append_binding_identifiers(idents);
+                }
+            }
         }
     }
 
```

connorshea committed:
feat(linter): Enable commented-out test for prefer-const rule.

This was fixed by #15710

connorshea committed:
Resolve feedback.

connorshea committed:
Make comment clearer.

connorshea committed:
refactor(linter): Simplify the code for prefer-const rule implementation.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

connorshea committed:
Clean up code a bit more.

Resolve some of the most deeply nested if statements here so the code is less difficult to read through.

connorshea committed:
docs(linter): Improve docs for prefer-const rule.

connorshea committed:
Update the prefer-const config to deserialize using the config object.

connorshea committed:
Add basic tests for ensuring comments don't get caught.

connorshea committed:
Update the diagnostic to use a period at the end of sentences.

Also use backticks for code references.

connorshea committed:
Fix prefer-const to not suggest const for cases where it is invalid.

`let foo;` is valid, but `const foo;` is not. So this kind of cases should be a violation.

connorshea committed:
Fix prefer-const to not suggest const for another case where it is invalid.

Add a test for this case so it doesn't regress.

connorshea committed:
Add one more test case for commented code.

Just to be sure that we don't misinterpret commented out code as actual code.

connorshea committed:
fix(linter): Fix another edge-case for a case where const cannot be used.

inside for statements, it's not really practical to change let to const like this.

Also add some basic cases for typescript code.

connorshea committed:
formatting tweaks for tests

camc314 commented on crates/oxc_linter/src/rules/eslint/prefer_const.rs:
```diff
@@ -0,0 +1,1062 @@
+
+            if write_only_refs.len() == 1 {
+                let write_ref = write_only_refs[0];
+                let write_node_id = write_ref.node_id();
```
```suggestion

            if let Some(write_ref) = write_only_refs.next()
                && write_only_refs.next().is_none()
            {
                let write_node_id = write_ref.node_id();
```

Avoid collecting here, as collecting introduces a memory allocation which is slow

camc314 commented on crates/oxc_linter/src/rules/eslint/prefer_const.rs:
```diff
@@ -0,0 +1,1062 @@
+                    let read_span = ctx.nodes().get_node(r.node_id()).kind().span();
+                    let write_span = ctx.nodes().get_node(write_node_id).kind().span();
+                    read_span.start < write_span.start
+                });
```
how does eslint do this. Relying on spans will be unreliable as functions are hoisted ect.

connorshea committed:
Resolve feedback on memory usage.

Avoid using `collect()` when unnecessary while checking for write-only refs.

connorshea committed:
refactor: Rework the ignoreReadBeforeAssign logic.

Comparing span starts doesn't necessarily work, because of hoisting, so compare nodes instead to rely on the semantic order.

connorshea committed:
Simplify a bit and add tests.

connorshea committed:
Add comment
